hs_core/hydroshare/resource.py

```diff
@@ -1001,15 +1001,25 @@ def delete_resource_file(pk, filename_or_id, user, delete_logical_file=True):
     resource = utils.get_resource_by_shortkey(pk)
     res_cls = resource.__class__
 
+    def can_delete_logical_file(logical_file):
+        if not logical_file.is_fileset and not logical_file.is_model_program and not logical_file.is_model_instance:
+            # delete logical file if any resource file that belongs to logical file
+            # gets deleted for any logical file other than fileset or modelprogram or modelinstance logical file
+            return True
+        elif (logical_file.is_model_program or logical_file.is_model_instance) and logical_file.folder is None:
+            # this is a single file based model program/instance aggregation and not a folder based
+            # model program/instance aggregation - we need to delete the logical file
+            return True
+        return False
+
     for f in ResourceFile.objects.filter(object_id=resource.id):
```

Comment:
This could potentially loop through thousands of files with some resources.  Please add the filter_condition logic to the filter.